File: pa1/List.java
/*------------------------------------------------
File Name: List.java
Name:   Joseph Csoti
CruzID: 1617438 // jcsoti
Class:  CMPS 101
Date:   10/09/18
------------------------------------------------*/

class List {

  Node head;
  Node tail;
  Node cur;

  class Node {

    Node prev;
    Node next;
    int val;

    Node(int val) {
      this.val = val;
      this.next = null;
      this.prev = null;
    }

  }

  // Constructor
  List() {
    head = null;
    tail = null;
    cur = null;
  }

  // Returns the number of elements in this List.
  int length() {
    if (head == null)
      return 0;

    int len = 0;
    Node cur = head;

    while (cur != null) {
      len++;
      cur = cur.next;
    }

    return len;
  }

  // If cursor is defined, returns the index of the cursor element,
  // otherwise returns -1.
  int index() {
    if (this.cur == null)
      return -1;

    int len = 0;
    Node cur = head;

    while (cur != null) {
      if (this.cur == cur)
        return len;
      cur = cur.next;
      len++;
    }

    return len;
  }

  // Returns front element. Pre: length()>0
  int front() throws RuntimeException {
    if (length() <= 0)
      throw new RuntimeException("List.java: front() - List is empty");

    return head.val;
  }

  // Returns back element. Pre: length()>0
  int back() throws RuntimeException {
    if (length() <= 0)
      throw new RuntimeException("List.java: back() - List is empty");

    return tail.val;
  }

  // Returns cursor element. Pre: length()>0, index()>=0
  int get() throws RuntimeException {
    if (length() <= 0)
      throw new RuntimeException("List.java: get() - List is empty");
    if (index() < 0)
      throw new RuntimeException("List.java: get() - Invalid index");

    return cur.val;
  }

  // Returns true if and only if this List and L are the same
  // integer sequence. The states of the cursors in the two Lists
  // are not used in determining equality.
  boolean equals(List L) {

    // check size
    if (this.length() == L.length()) {

      Node cur_a = this.head;
      Node cur_b = L.head;
      // check items
      while (cur_a != null || cur_b != null) {
        if (cur_a.val != cur_b.val)
          return false; // Value @ curr is different
        cur_a = cur_a.next;
        cur_b = cur_b.next;
      }
      return true; // All items are equal

    } else
      return false; // Size is diff

  }

  // Resets this List to its original empty state.
  void clear() {
    head = null;
    tail = null;
    cur = null;
  }

  // If List is non-empty, places the cursor under the front element,
  // otherwise does nothing.
  void moveFront() {
    if (length() > 0)
      cur = head;
  }

  // If List is non-empty, places the cursor under the back element,
  // otherwise does nothing.
  void moveBack() {
    if (length() > 0)
      cur = tail;
  }

  // If cursor is defined and not at front, moves cursor one step toward
  // front of this List, if cursor is defined and at front, cursor becomes
  // undefined, if cursor is undefined does nothing.
  void movePrev() {
    if (cur == null)
      return;

    if (cur == head) {
      cur = null;
      return;
    }

    if (cur != head) {
      cur = cur.prev;
      return;
    }
  }

  // If cursor is defined and not at back, moves cursor one step toward
  // back of this List, if cursor is defined and at back, cursor becomes
  // undefined, if cursor is undefined does nothing.
  void moveNext() {
    if (cur == null)
      return;

    if (cur == tail) {
      cur = null;
      return;
    }

    if (cur != tail) {
      cur = cur.next;
      return;
    }

  }

  // Insert new element into this List. If List is non-empty,
  // insertion takes place before front element.
  void prepend(int data) {
    Node node = new Node(data);

    // First node
    if (head == null) {
      head = node;
      tail = node;
      return;
    }

    // nth node
    head.prev = node;
    node.next = head;
    head = node;

  }

  // Insert new element into this List. If List is non-empty,
  // insertion takes place after back element.
  void append(int data) {
    Node node = new Node(data);

    // First node
    if (head == null) {
      head = node;
      tail = node;
      return;
    }

    // nth node
    node.prev = tail;
    tail.next = node;
    tail = node;

  }

  // Insert new element before cursor.
  // Pre: length()>0, index()>=0
  void insertBefore(int data) throws RuntimeException {
    if (length() <= 0)
      throw new RuntimeException("List.java: insertBefore() - List is empty");
    if (index() < 0)
      throw new RuntimeException("List.java: insertBefore() - Invalid index");

    // Equivalent to prepending
    if (cur == head) {
      prepend(data);
      return;
    }

    // prepare node
    Node node = new Node(data);
    node.prev = cur.prev;
    node.next = cur;

    // Insert at cur
    cur.prev.next = node;
    cur.prev = node;

  }

  // Inserts new element after cursor.
  // Pre: length()>0, index()>=0
  void insertAfter(int data) throws RuntimeException {
    if (length() <= 0)
      throw new RuntimeException("List.java: insertAfter() - List is empty");
    if (index() < 0)
      throw new RuntimeException("List.java: insertAfter() - Invalid index");

    // Equivalent to appending
    if (cur == tail) {
      append(data);
      return;
    }

    // Prepare node
    Node node = new Node(data);
    node.prev = cur;
    node.next = cur.next;

    // Insert node
    cur.next = node;
    cur.next.prev = node;
  }

  // Deletes the front element. Pre: length()>0
  void deleteFront() throws RuntimeException {
    if (length() <= 0)
      throw new RuntimeException("List.java: deleteFront() - List is empty");

    removeNode(head);
  }

  // Deletes the back element. Pre: length()>0
  void deleteBack() throws RuntimeException {
    if (length() <= 0)
      throw new RuntimeException("List.java: deleteBack() - List is empty");

    removeNode(tail);
  }

  // Deletes cursor element, making cursor undefined.
  // Pre: length()>0, index()>=0
  // Other methods
  void delete() throws RuntimeException {
    if (length() <= 0)
      throw new RuntimeException("List.java: delete() - List is empty");
    if (index() < 0)
      throw new RuntimeException("List.java: delete() - Invalid index");

    removeNode(cur);
  }

  // Overrides Object's toString method. Returns a String
  // representation of this List consisting of a space
  // separated sequence of integers, with front on left.
  public String toString() {
    String res = "";

    Node cur = head;
    while (cur != null) {
      res += cur.val + " ";
      cur = cur.next;
    }

    return res;
  }

  // Returns a new List representing the same integer sequence as this
  // List. The cursor in the new list is undefined, regardless of the
  // state of the cursor in this List. This List is unchanged.
  List copy() {

    List copy = new List();
    Node cur = head;

    while (cur != null) {
      copy.append(cur.val);
      cur = cur.next;
    }

    return copy;
  }

  // Generic helper function to help remove an given node
  void removeNode(Node node) {

    // Node n is cursor so set cur to (-1)
    if (cur == node)
      this.cur = null;

    // Only one node in the list
    if (length() <= 1) {
      head = null;
      tail = null;
      return;
    }

    // Remove head
    if (node == head) {
      head.next.prev = null;
      head = head.next;
      return;
    }

    // Remove tail
    if (node == tail) {
      tail.prev.next = null;
      tail = tail.prev;
      return;
    }

    // Remove body
    Node cur = head;
    while (cur.next != null) {
      if (cur == node) {
        cur.prev.next = cur.next;
        cur.next.prev = cur.prev;
        return;
      }
      cur = cur.next;
    }

  }

}

File: pa2/List.c
/*------------------------------------------------
File Name: List.c
Name:   Joseph Csoti
CruzID: 1617438 // jcsoti
Class:  CMPS 101
Date:   10/22/18
------------------------------------------------*/
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

#include "List.h"

// private types --------------------------------

// NodeObj
typedef struct NodeObj{
  int* item;
  struct NodeObj* next;
  struct NodeObj* prev;
} NodeObj;

// newNode()
// constructor of the Node type
Node newNode(int* item) {
  Node N = malloc(sizeof(NodeObj));
  assert(N!=NULL);
  N->item = item;
  N->prev = NULL;
  N->next = NULL;
  return(N);
}

// freeNode()
// destructor for the Node type
void freeNode(Node pN){
  if(pN != NULL) {
    free(pN);
    pN = NULL;
  }
}
// ------------------------------------

//Listobj
typedef struct ListObj {
  struct NodeObj* head;
  struct NodeObj* tail;
  struct NodeObj* cursor;
} ListObj;

// Constructors-Destructors ---------------------------------------------------
List newList(void) {
  List L = malloc(sizeof(ListObj));
  assert(L!=NULL);
  return L;
}

void freeList(List* pL) {
  if( pL!=NULL && *pL!=NULL ){
    free(*pL);
    *pL = NULL;
 }
}

// Access functions -----------------------------------------------------------
int length(List L) {

  if(L->head == NULL)
    return 0;
  if(L->head == L->tail)
    return 1;

  int len = 0;
  Node cur = L->head;

  while(cur != NULL) {
    len++;
    cur = cur->next;
  }

  return len;
}

int index(List L) {
  if(L->cursor == NULL)
    return -1;

  int index = 0;
  Node cur = L->head;

  while(cur != NULL){
    if(L->cursor == cur)
      return index;
    
    cur = cur->next;
    index++;
  }

  return index;
}

int front(List L) {
  if(length(L) <= 0) {
    fprintf(stderr, "List.c: front() called on an empty list\n");
    exit(1);
  }
  return *(L->head->item);
}

int back(List L) {
  if(length(L) <= 0) {
    fprintf(stderr, "List.c: back() called on an empty list\n");
    exit(1);
  }
  return *(L->tail->item);
}

int get(List L) {
  if(length(L) <= 0) {
    fprintf(stderr, "List.c: get() called on an empty list\n");
    exit(1);
  }
  if(index(L) < 0) {
    fprintf(stderr, "List.c: get() invalid index\n");
    exit(1);
  }
  return *(L->cursor->item);
}

int equals(List A, List B) {
  // check size
  if (length(A) == length(B)) {
    Node cur_a = A->head;
    Node cur_b = B->head;
    // check items
    while (cur_a != NULL || cur_b != NULL) {
      if (cur_a->item != cur_b->item)
        return 0; // Value @ curr is different
      cur_a = cur_a->next;
      cur_b = cur_b->next;
    }
      return 1; // All items are equal
    } else return 0; // Size is diff

    return 0;
}

// Manipulation procedures ----------------------------------------------------
void clear(List L) {

Node curr = L->head;
while( curr != NULL ) {
   Node next = curr->next;
   free(curr->item)
   free(curr);
   curr = next;
}
  //freeNode(L->cursor);
}
void moveFront(List L) {
  if(length(L) > 0)
    L->cursor = L->head;
}
void moveBack(List L){
  if(length(L) > 0)
    L->cursor = L->tail;
}
void movePrev(List L){
  if(L->cursor == NULL)
    return;
  
  if (L->cursor == L->head) {
    freeNode(L->cursor);
    return;
  }

  if (L->cursor != L->head) {
    L->cursor = L->cursor->prev;
    return;
  }
}
void moveNext(List L) {
  if (L->cursor == NULL)
    return;

  if (L->cursor == L->tail) {
    freeNode(L->cursor);
    return;
  }

  if (L->cursor != L->tail) {
    L->cursor = L->cursor->next;
    return;
  }
}
void prepend(List L, int data){
    Node node = newNode(&data);

    // First node
    if (L->head == NULL) {
      L->head = node;
      L->tail = node;
      return;
    }

    // nth node
    L->head->prev = node;
    node->next = L->head;
    L->head = node;
}
void append(List L, int data){
  Node node = newNode(&data);

   // First node
  if (L->head == NULL) {
    L->head = node;
    L->tail = node;
    return;
  }

  // nth node
  node->prev = L->tail;
  L->tail->next = node;
  L->tail = node;
}
void insertBefore(List L, int data){
    if (length(L) <= 0) {
      fprintf(stderr, "List.c: insertBefore() called on an empty list\n");
      return;
    }
    if (index(L) < 0) {
      fprintf(stderr, "List.c: insertBefore() invalid index\n");
      return;
    }

    // Equivalent to prepending
    if (L->cursor == L->head) {
      prepend(L, data);
      return;
    }

    // prepare node
    Node node = newNode(&data);
    node->prev = L->cursor->prev;
    node->next = L->cursor;

    // Insert at cur
    L->cursor->prev->next = node;
    L->cursor->prev = node;
}
void insertAfter(List L, int data){
    if (length(L) <= 0) {
      fprintf(stderr, "List.c: insertAfter() called on an empty list\n");
      return;
    }
    if (index(L) < 0) {
      fprintf(stderr, "List.c: insertAfter() invalid index\n");
      return;
    }

      // Equivalent to appending
    if (L->cursor == L->tail) {
      append(L, data);
      return;
    }

    // Prepare node
    Node node = newNode(&data);
    node->prev = L->cursor;
    node->next = L->cursor->next;

    // Insert node
    L->cursor->next = node;
    L->cursor->next->prev = node;
}
void deleteFront(List L){
    if (length(L) <= 0) {
      fprintf(stderr, "List.c: deleteFront() called on an empty list\n");
      return;
    }

    removeNode(L, L->head);
}
void deleteBack(List L){
    if (length(L) <= 0) {
      fprintf(stderr, "List.c: deleteBack() called on an empty list\n");
      return;
    }

    removeNode(L, L->tail);
}
void delete(List L){
    if (length(L) <= 0) {
      fprintf(stderr, "List.c: delete() called on an empty list\n");
      return;
    }
    if (index(L) < 0) {
      fprintf(stderr, "List.c: delete() invalid index\n");
      return;
    }

    removeNode(L, L->cursor);
}
// Custom node remover
void removeNode(List L, Node node) {
      // Node n is cursor so set cur to (-1)
    if (L->cursor == node)
      freeNode(L->cursor);
    // buffer


    // Only one node in the list
    if (length(L) == 1) {
      freeNode(L->head);
      freeNode(L->tail);
      return;
    }

    // Remove head
    if (node == L->head) {
      freeNode(L->head->next->prev);
      L->head = L->head->next;
      return;
    }

    // Remove tail
    if (node == L->tail) {
      freeNode(L->tail->prev->next);
      L->tail = L->tail->prev;
      return;
    }

    // Remove body
    Node cur = L->head;
    while (cur->next != NULL) {
      if (cur == node) {
        cur->prev->next = cur->next;
        cur->next->prev = cur->prev;
        return;
      }
      cur = cur->next;
    }
}

// Other operations -----------------------------------------------------------
void printList(FILE* out, List L){
    Node cur = L->head;
    while (cur != NULL) {
      fprintf(out, "%d ", *(cur->item) );
      cur = cur->next;
    }

}
List copyList(List L){
    List copy = newList();
    Node cur = L->head;

    while (cur != NULL) {
      append(copy, *(cur->item));
      cur = cur->next;
    }

    return copy;
}